File: src/main/java/com/github/svegon/capi/event/entity/EntityHorizontalSpeedMultiplierCallback.java
package com.github.svegon.capi.event.entity;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.entity.Entity;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

public interface EntityHorizontalSpeedMultiplierCallback {
    Event<EntityHorizontalSpeedMultiplierCallback> EVENT =
            EventFactory.createArrayBacked(EntityHorizontalSpeedMultiplierCallback.class,
            (entity, cir) -> {}, listeners -> (entity, cir) -> {
        for (EntityHorizontalSpeedMultiplierCallback listener : listeners) {
            listener.getEntityHorizontalSpeedMultiplier(entity, cir);

            if (cir.isCancelled()) {
                return;
            }
        }
    });

    void getEntityHorizontalSpeedMultiplier(Entity entity, CallbackInfoReturnable<Float> cir);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/entity/EntityHorizontalSpeedMultiplierCallback.kt
package io.github.svegon.mclientapi.event.entity

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.entity.Entity
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable
import java.util.function.Function
import kotlin.Array
import kotlin.Float
import kotlin.invoke

interface EntityHorizontalSpeedMultiplierCallback {
    fun getEntityHorizontalSpeedMultiplier(entity: Entity?, cir: CallbackInfoReturnable<Float?>?)

    companion object {
        @JvmField
        val EVENT: Event<EntityHorizontalSpeedMultiplierCallback?> = EventFactory.createArrayBacked(
            EntityHorizontalSpeedMultiplierCallback::class.java,
            EntityHorizontalSpeedMultiplierCallback { entity: Entity?, cir: CallbackInfoReturnable<Float?>? -> },
            Function { listeners: Array<EntityHorizontalSpeedMultiplierCallback?>? ->
                EntityHorizontalSpeedMultiplierCallback { entity: Entity?, cir: CallbackInfoReturnable<Float?>? ->
                    for (listener in listeners) {
                        listener.getEntityHorizontalSpeedMultiplier(entity, cir)

                        if (cir.isCancelled()) {
                            return@EntityHorizontalSpeedMultiplierCallback
                        }
                    }
                }
            })
    }
}
